scio-tensorflow/src/main/scala/com/spotify/scio/tensorflow/TensorFlowFunctions.scala

```diff
@@ -263,12 +272,112 @@ class TFExampleSCollectionFunctions[T <: Example](val self: SCollection[T]) {
+  : (Future[Tap[Example]], Future[Tap[Schema]]) = {
+    val inferedSchema = Option(schema).getOrElse(self.inferExampleMetadata(ev))
+    val schemaPath = path.replaceAll("\\/+$", "") + "/" + schemaFilename
+    if (self.context.isTest) {
+      self.context.testOut(TFExampleIO(path))(self.asInstanceOf[SCollection[Example]])
+      self.context.testOut(ProtobufIO(schemaPath))(inferedSchema)
+      (self.saveAsInMemoryTap.asInstanceOf[Future[Tap[Example]]], inferedSchema.saveAsInMemoryTap)
+    } else {
+      import scala.concurrent.ExecutionContext.Implicits.global
+      saveExampleMetadata(inferedSchema, schemaPath)
+      val r = self.map(_.toByteArray).saveAsTfRecordFile(path, suffix, compression, numShards)
+      (r.map(_.map(Example.parseFrom)), Taps().protobufFile[Schema](schemaPath))
+    }
+  }
+
+  /**
+   * Infer a `org.tensorflow.metadata.v0.Schema` from this SCollection of
+   * `org.tensorflow.example.Example`.
+   * @return A singleton `SCollection` containing the schema.
+   */
+  def inferExampleMetadata(implicit ev: T <:< Example): SCollection[Schema] =
+    examplesToFeatures(self.asInstanceOf[SCollection[Example]])
+      .groupBy(_ => ())
+      .values.map(features => Schema.newBuilder().addAllFeature(features.asJava).build())
+
+  private def saveExampleMetadata(schema: SCollection[Schema], schemaPath: String)
+  : SCollection[Schema] = {
+    schema.map { s =>
+      val d = FileSystems.matchNewResource(schemaPath, false)
+      val chnnl = Channels.newOutputStream(FileSystems.create(d, MimeTypes.BINARY))
+      try {
+        s.writeTo(chnnl)
+      } finally {
+        chnnl.close()
+      }
+    }
+    schema
+  }
+
+  private def examplesToFeatures(examples: SCollection[Example]): SCollection[Feature] =
+    examples
+      .flatMap(_.getFeatures.getFeatureMap.asScala)
+      .map { case (name, feature) =>
+        feature.getKindCase match {
+          case KindCase.BYTES_LIST =>
+            ((name, FeatureType.BYTES), feature.getBytesList.getValueCount)
+          case KindCase.FLOAT_LIST =>
+            ((name, FeatureType.FLOAT), feature.getFloatList.getValueCount)
+          case KindCase.INT64_LIST =>
+            ((name, FeatureType.INT), feature.getInt64List.getValueCount)
+          case KindCase.KIND_NOT_SET =>
+            sys.error(s"kind must be set - feature is ${feature.toString}")
+        }
+      }
+      .aggregateByKey(MultiAggregator((Aggregator.max[Int], Aggregator.min[Int])))
+      .map { case ((featureName, featureType), (max, min)) =>
+        val builder = Feature.newBuilder()
+          .setName(featureName)
+          .setType(featureType)
+        if (max == min) {
+          // This is a fixed length feature
+          builder.setShape(FixedShape.newBuilder()
+            .addDim(FixedShape.Dim.newBuilder().setSize(max)))
+        }
+        else {
+          // Var length
+          builder.setValueCount(ValueCount.newBuilder().setMin(min).setMax(max))
```

Comment:
Is this how these fields are supposed to be used?
what happens in the case where the max len of a var len field in the training set is N and when doing predictions, a new record comes with length > N? I assume these schema fields are not automatically used, it's up to the users to use it

Reply:
@yonromai this schema is used only for preprocessing (featran or tf.transform), which themselves would produce another version of schema for inference time, which could or could not include this info - probably not. does that make sense.